ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateRegenerationBaseSelector.java

```diff
@@ -0,0 +1,126 @@
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class StateRegenerationBaseSelector {
+  private final Optional<SlotAndBlockRoot> latestEpochBoundary;
+  private final Supplier<Optional<BlockRootAndState>> closestAvailableStateSupplier;
+  private final StateAndBlockProvider stateAndBlockProvider;
+  private final BlockProvider blockProvider;
+  private final Optional<SignedBlockAndState> rebasedStartingPoint;
+
+  public StateRegenerationBaseSelector(
+      final Optional<SlotAndBlockRoot> latestEpochBoundary,
+      final Supplier<Optional<BlockRootAndState>> closestAvailableStateSupplier,
+      final StateAndBlockProvider stateAndBlockProvider,
+      final BlockProvider blockProvider,
+      final Optional<SignedBlockAndState> rebasedStartingPoint) {
+    this.latestEpochBoundary = latestEpochBoundary;
+    this.closestAvailableStateSupplier = closestAvailableStateSupplier;
+    this.stateAndBlockProvider = stateAndBlockProvider;
+    this.blockProvider = blockProvider;
+    this.rebasedStartingPoint = rebasedStartingPoint;
+  }
+
+  public StateRegenerationBaseSelector withRebasedStartingPoint(
+      final SignedBlockAndState blockAndState) {
+    if (isBetterThanCurrentRebasedStartingPoint(blockAndState)
+        && isEqualToOrBetterThanLatestEpochBoundary(blockAndState)) {
+      return new StateRegenerationBaseSelector(
+          latestEpochBoundary,
+          closestAvailableStateSupplier,
+          stateAndBlockProvider,
+          blockProvider,
+          Optional.of(blockAndState));
+    }
+    return this;
+  }
+
+  private boolean isEqualToOrBetterThanLatestEpochBoundary(
+      final SignedBlockAndState blockAndState) {
+    return latestEpochBoundary.isEmpty()
+        || blockAndState.getSlot().isGreaterThanOrEqualTo(latestEpochBoundary.get().getSlot());
+  }
+
+  private boolean isBetterThanCurrentRebasedStartingPoint(final SignedBlockAndState blockAndState) {
+    return isBetterThan(
+        blockAndState.getSlot(), rebasedStartingPoint.map(SignedBlockAndState::getSlot));
+  }
+
+  public SafeFuture<Optional<SignedBlockAndState>> getBestBase() {
+    final Optional<BlockRootAndState> closestAvailableFromStore =
+        closestAvailableStateSupplier.get();
+    if (closestAvailableFromStore.isEmpty()) {
+      // Can't be a valid target state or has since been finalized. No point regenerating.
+      return SafeFuture.completedFuture(Optional.empty());
+    }
+
+    final Optional<UInt64> storeSlot = closestAvailableFromStore.map(BlockRootAndState::getSlot);
+    final Optional<UInt64> epochBoundarySlot = latestEpochBoundary.map(SlotAndBlockRoot::getSlot);
+    final Optional<UInt64> rebasedSlot = rebasedStartingPoint.map(SignedBlockAndState::getSlot);
+
+    if (epochBoundarySlot.isPresent()
+        && isBestOption(epochBoundarySlot.get(), storeSlot, rebasedSlot)) {
+      return stateAndBlockProvider
+          .getBlockAndState(latestEpochBoundary.get().getBlockRoot())
```

Comment:
Should we have a threshold where we only pull the epoch boundary state if it's significantly closer than `closestAvailableFromStore` and `rebasedStartingPoint`?

Reply:
hmm, yeah probably.  Will do in a separate PR though.